File: Lab 4/Java/Ass4up.java
import java.util.*;

public class Ass4up {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt())
            return;
        int t = sc.nextInt();
        while (t-- > 0) {
            if (!sc.hasNext())
                break;

            // Input format in Python: "temp_var = input().split(); vertices =
            // int(temp_var[0])..."
            // Java scanner handles whitespace.
            int n = sc.nextInt();
            int m = sc.nextInt();

            int[] starts = new int[m];
            for (int i = 0; i < m; i++)
                starts[i] = sc.nextInt();

            int[] ends = new int[m];
            for (int i = 0; i < m; i++)
                ends[i] = sc.nextInt();

            int[] degree = new int[n];
            for (int i = 0; i < m; i++) {
                degree[starts[i] - 1]++;
                degree[ends[i] - 1]++;
            }

            int odd = 0;
            for (int d : degree) {
                if (d % 2 != 0)
                    odd++;
            }

            if (odd == 0 || odd == 2)
                System.out.println("YES");
            else
                System.out.println("NO");
        }
    }
}

File: Lab 4/Ass4up.py
# PROBLEM: Determine if an Eulerian path or circuit exists in undirected graph
# INPUT: Multiple test cases, each with vertices, edges, and two arrays (starts, ends)
# OUTPUT: "YES" if graph has Eulerian path/circuit (0 or 2 odd-degree vertices), "NO" otherwise
# KEYWORDS: Euler path, Euler circuit, graph theory, degree counting, undirected graph

import sys  # fast stdin

input = sys.stdin.readline  # quick line reads

# read number of test cases
t_var = int(input())

# process each test case
while t_var > 0:

    temp_var = input().split()  # first line tokens
    
    vertices_var = int(temp_var[0])  # number of vertices
    
    edges_var = int(temp_var[1])  # number of edges
    
    
    start_var = list(map(int, input().split()))  # edge starts (1-based)
    end_var = list(map(int, input().split()))  # edge ends (1-based)
    
    
    degree_var = [0] * vertices_var  # degree counts per vertex (0-based indexing)
    
    
    i_var = 0
    while i_var < edges_var:  # accumulate degrees
        
        u_var = start_var[i_var] - 1  # convert to 0-based
        v_var = end_var[i_var] - 1
        
        degree_var[u_var] += 1  # out/in both count for undirected degree
    
        degree_var[v_var] += 1
        
        i_var = i_var + 1
    
    
    odd_var = 0  # count of odd-degree vertices
    i_var = 0
    
    while i_var < vertices_var:
        
        if degree_var[i_var] % 2 != 0:  # odd degree check
    
            odd_var = odd_var + 1
        
        i_var = i_var + 1
    
    
    if odd_var == 0 or odd_var == 2:  # Euler circuit/path condition
    
        print("YES")
    else:
        print("NO")
    
    # move to next test case
    t_var = t_var - 1
